### DISCUSSION TITLE
Non-continuous jog movement

### DISCUSSION THREAD

**Discussion by mvackel** [Paciente8159/uCNC#498] [Config and issues] [answered]
(uCNC 1.7.4  using Mega/RAMPS 1.4 5-axis)

After initial boot, the jog movement is not continuous on any axis, moving in noticeable steps. For instance, if you jog to move 5 millimeter (using UGS or the controller), the motor move in about 20 steps and not continuously. The axes are all configured for 400 steps/mm.

In UGS, if you send a Soft Reset command, all axis start moving continuously, fixing the problem.

If you go to the controller (Reprapdiscount full graphic smart display) and issue a jog command from there on any axis, the jerky movement returns and is again observed issuing the command by UGS.

Any way to correct this?

**Comment by Paciente8159** [comment]
The button motion is also made to cancel jog motion. Maybe this is causing issues.
For example if you send a jog command and then press the button mid jog it should stop it I think. I got to review the code to be sure.

**Comment by Paciente8159** [comment]
About the making the jog motion more continuous there is a change that can be made.
Currently the system menu only accepted one jog command at time.
I will modify that to allow chaining multiple jog commands like it's possible to to with UGS.

About the problems of jog consistency I will still test that to see if I can replicate the problem by using both the display and UGS at the same time.

**Comment by Paciente8159** [comment]
#501 should allow multiple JOG commands to be chained via the display controller.
The issue of the very slow JOG happened once while using the display but I was not unable to determine the source. (I'm still investigating).
After I sent a JOG cancel command the JOG it started to work ok again.

**Reply by mvackel** [reply]
The last modification broke the workaround of issuing a "Soft Reset" from UGS to the board to make the axis movement continuous. Now, even after a Soft Reset (from UGS) the movement is non-continuous. This is a problem as I can't use manually the machine.
The Soft Reset from the display (5 secs button press) also made no difference in the axes' movements.
Correcting.... after a few board resets, if I send a Soft Reset just before EACH jog movement, the movement is continuous. The second jog returns to the hiccup movement.

**Reply by Paciente8159** [reply]
Ok thanks. I will investigate. The changes to the core code was minimal and only affected system menu module.

One question. If you compile a run code on UGS with the display modules disabled do you have any issue with jogging?

**Reply by mvackel** [reply]
Good question. I've just tested and yes, the culprit is the graphic display module. 
I tested:
- Completely removed -> Jogs Ok
- +lib_ignore + u8g2   -> Jogs Ok
- +Enable main loop     -> jogs Ok
- +Enable graphic display  -> Jogs in pulses like hiccups.

**Reply by Paciente8159** [reply]
Thanks.
I will see if I can reproduce the issue on my RAMPS with the display and UGS. I'll give you feedback as soon as I find something.

**Reply by Paciente8159** [reply]
One question?

Does this issue affects only jog? What about normal motions or running a gcode file?

I do get some kind of jagged motion but it seems to me that this is being caused by the idle screen refresh rate.

The idle screen refresh takes some processing time that leads the planner/interpolator buffer to get starved and to constantly accel/deaccel as new motion data is feed.
This should also manifest on normal gcode execution too.

**EDIT**

Try to add this line to your cnc_hal_override.h file and see if it improves the issue
`#define SYSTEM_MENU_REDRAW_IDLE_MS 2000`

**Reply by mvackel** [reply]
I tried issuing G0 and G1 and the movement is jagged. I don't know on a program because the movement is usually mixed with many other commands.
I tried to add the line you suggested and the movement is MUCH better, but still interrupted, apparently in intervals of 2 seconds.

**Reply by Paciente8159** [reply]
Then (that confirms my observation) the issue is the toll the screen update takes in the main loop. The motion planner is being starved by the time it takes to draw the full screen.
I'm reading thought the u8g2 to try to improve the update speed of the screen or even do partial updates.

I'll give you some feedback when I come up with an improved module.

**Reply by Paciente8159** [reply, marked as answer]
I believe this fixes the issue.
https://github.com/Paciente8159/uCNC-modules/pull/33

All u8g2 function call take along time to execute and if too many of them are called without calling the main loop the planner gets empty and the motion gets jagged.
This also enforces that the main loop is called while communicating with the LCD.

After making these modifications the motion seems to be uniform even navigating inside system menus.

Please let me know if you find any issue.

**Reply by mvackel** [reply]
The movement is fixed using both: the display and UGS. Great!

I'm not sure if this is just after this modification, but the jog control using the display's rotary encoder got very unresponsive and/or erratic again. I have to command (turn the encoder) about 10 times to get just one or two responses and some of them in the wrong direction. If I rotate the encoder VERY slowly (so that the movement occurs before each "click"), it works, but sometimes in the wrong direction. The menu item navigation is ok, as you mentioned - as is the change of numeric values, like "Settings/Axis/X-accel".  

Should I open another discussion about this?

**Reply by Paciente8159** [reply]
I think we can continue tracking the issue [here](https://github.com/Paciente8159/uCNC/discussions/497)

**Reply by Paciente8159** [reply]
And clarifying. The issue you are experiencing is also a bit a of a consequence of the modification.
I'll explain what is happening.

The encoder reading and screen refresh are part of the main loop, since they are hooked via extension events.
The problem I found is that the redrawing of the screen was taking too long to perform and would lead to the step interpolator getting empty and causing a jagged motion.

This forced me to recall the main loop again (several times and from several places) from the module to keep feeding the interpolator and motion planner.

To prevent the "inception problem" and to blow up the stack, once inside graphic display code the code as to prevent calling it self.

This also affect the amount of times the encoder is read.
I have a new idea how to cope with this and will try to test this later today.

I will pull the encoder reading outside and add an accumulator that will count the number of pulses and direction.
It will then execute the use action based on the accumulated value instead of the pulse reading at that moment.

This should hopefully improve wrong direction (it will not fixed it completely but hopefully improve it), and will also help by acting as a buffer of the user action in between screen refreshing frames.

I'll let you know if it works and when it's done.
